File: ai-service/app/services/recommendation_service.py
from __future__ import annotations

from collections import Counter
from datetime import datetime
from pathlib import Path
from typing import Any

import pandas as pd

from app.models.recommendation_model import BookingRecommender

REQUIRED_COLUMNS = ["user_id", "court_id", "day_of_week", "time_slot"]
MAX_RECOMMENDATIONS = 3


def _normalize_time(value: Any) -> str:
    raw = str(value or "").strip()
    if len(raw) >= 5:
        return raw[:5]
    return raw


def _build_time_label(start_time: str, end_time: str) -> str:
    return f"{_normalize_time(start_time)} - {_normalize_time(end_time)}"


def _build_time_key(start_time: str, end_time: str) -> str:
    return f"{_normalize_time(start_time)}|{_normalize_time(end_time)}"


def _build_court_time_key(court_id: int, start_time: str, end_time: str) -> str:
    return f"{int(court_id)}|{_build_time_key(start_time, end_time)}"


def _normalized_counter_value(counter: Counter, key: Any) -> float:
    if not counter:
        return 0.0

    max_value = max(counter.values())
    if max_value <= 0:
        return 0.0

    return counter.get(key, 0) / max_value


def _clamp_score(value: float) -> float:
    return max(0.0, min(1.0, float(value)))


def _get_day_of_week(date_string: str) -> str:
    return datetime.strptime(date_string, "%Y-%m-%d").strftime("%A")


def _score_prime_time(start_time: str) -> float:
    try:
        hour = int(_normalize_time(start_time)[:2])
    except ValueError:
        return 0.4

    return _clamp_score(1 - min(abs(hour - 19) / 12, 1))


class RecommendationService:
    def __init__(self, dataset_path: Path) -> None:
        self.dataset_path = dataset_path
        self.model = BookingRecommender()
        self.global_time_counts: Counter[str] = Counter()
        self.global_day_time_counts: Counter[tuple[str, str]] = Counter()
        self.global_court_time_counts: Counter[tuple[int, str]] = Counter()
        self._initialize_model()

    def _initialize_model(self) -> None:
        if not self.dataset_path.exists():
            raise RuntimeError(f"Dataset not found: {self.dataset_path}")

        dataset = pd.read_csv(self.dataset_path)
        missing = [column for column in REQUIRED_COLUMNS if column not in dataset.columns]
        if missing:
            raise RuntimeError(f"Dataset missing required columns: {missing}")

        clean_dataset = dataset[REQUIRED_COLUMNS].dropna().copy()
        clean_dataset["user_id"] = clean_dataset["user_id"].astype(int)
        clean_dataset["court_id"] = clean_dataset["court_id"].astype(int)
        clean_dataset["day_of_week"] = clean_dataset["day_of_week"].astype(str)
        clean_dataset["time_slot"] = clean_dataset["time_slot"].astype(str).map(_normalize_time)

        self.model.fit(clean_dataset)
        self.global_time_counts = Counter(clean_dataset["time_slot"].tolist())
        self.global_day_time_counts = Counter(zip(clean_dataset["day_of_week"], clean_dataset["time_slot"]))
        self.global_court_time_counts = Counter(zip(clean_dataset["court_id"], clean_dataset["time_slot"]))

    def recommend_best_time(self, user_id: int) -> list[str]:
        if user_id <= 0:
            raise ValueError("user_id must be a positive integer")
        return self.model.recommend_best_time(user_id=user_id, top_n=MAX_RECOMMENDATIONS)

    def score_recommendations(self, payload: dict[str, Any]) -> dict[str, Any]:
        available_slots = [self._normalize_slot(slot) for slot in payload.get("availableSlots", [])]
        history = [self._normalize_history(item) for item in payload.get("history", [])]

        if not available_slots:
            return {
                "strategy": "empty",
                "recommendedOptions": [],
            }

        target_date = str(payload.get("targetDate") or "").strip()
        target_day = _get_day_of_week(target_date)
        history_stats = self._build_history_stats(history)
        has_history = bool(history)

        ranked_options = [
            self._score_slot(
                slot=slot,
                available_slots=available_slots,
                history_stats=history_stats,
                target_day=target_day,
                has_history=has_history,
            )
            for slot in available_slots
        ]

        ranked_options.sort(
            key=lambda item: (-item["score"], item["priceVnd"], item["courtName"], item["startTime"])
        )

        return {
            "strategy": "personalized" if has_history else "cold_start",
            "recommendedOptions": ranked_options[:MAX_RECOMMENDATIONS],
        }

    def _normalize_slot(self, slot: dict[str, Any]) -> dict[str, Any]:
        return {
            "courtId": int(slot["courtId"]),
            "courtName": str(slot.get("courtName") or "Sân"),
            "location": str(slot.get("location") or ""),
            "slotId": int(slot["slotId"]),
            "label": _build_time_label(slot.get("startTime"), slot.get("endTime")),
            "startTime": _normalize_time(slot.get("startTime")),
            "endTime": _normalize_time(slot.get("endTime")),
            "priceVnd": float(slot.get("priceVnd") or 0),
        }

    def _normalize_history(self, item: dict[str, Any]) -> dict[str, Any]:
        booking_date = str(item.get("bookingDate") or "").strip()
        day_of_week = str(item.get("dayOfWeek") or "").strip()
        if not day_of_week and booking_date:
            day_of_week = _get_day_of_week(booking_date)

        return {
            "bookingDate": booking_date,
            "dayOfWeek": day_of_week,
            "courtId": int(item["courtId"]),
            "courtName": str(item.get("courtName") or "Sân"),
            "startTime": _normalize_time(item.get("startTime")),
            "endTime": _normalize_time(item.get("endTime")),
            "priceVnd": float(item.get("priceVnd") or 0),
        }

    def _build_history_stats(self, history: list[dict[str, Any]]) -> dict[str, Any]:
        court_time_counts: Counter[str] = Counter()
        court_counts: Counter[int] = Counter()
        time_counts: Counter[str] = Counter()
        day_counts: Counter[str] = Counter()
        prices: list[float] = []

        for item in history:
            court_time_counts[_build_court_time_key(item["courtId"], item["startTime"], item["endTime"])] += 1
            court_counts[item["courtId"]] += 1
            time_counts[_build_time_key(item["startTime"], item["endTime"])] += 1
            day_counts[item["dayOfWeek"]] += 1
            prices.append(float(item["priceVnd"]))

        average_price = sum(prices) / len(prices) if prices else None
        return {
            "court_time_counts": court_time_counts,
            "court_counts": court_counts,
            "time_counts": time_counts,
            "day_counts": day_counts,
            "average_price": average_price,
        }

    def _score_price_fit(
        self,
        price_vnd: float,
        average_price: float | None,
        available_slots: list[dict[str, Any]],
    ) -> float:
        if average_price and average_price > 0:
            return _clamp_score(1 - min(abs(price_vnd - average_price) / average_price, 1))

        prices = [float(slot["priceVnd"]) for slot in available_slots]
        max_price = max(prices)
        min_price = min(prices)
        if max_price == min_price:
            return 0.7

        return _clamp_score(1 - (price_vnd - min_price) / (max_price - min_price))

    def _score_global_signal(self, slot: dict[str, Any], target_day: str) -> float:
        start_time = slot["startTime"]
        court_id = int(slot["courtId"])
        time_score = _normalized_counter_value(self.global_time_counts, start_time)
        day_time_score = _normalized_counter_value(self.global_day_time_counts, (target_day, start_time))
        court_time_score = _normalized_counter_value(self.global_court_time_counts, (court_id, start_time))
        return (time_score + day_time_score + court_time_score) / 3

    def _build_reason(
        self,
        slot: dict[str, Any],
        *,
        has_history: bool,
        court_time_score: float,
        court_score: float,
        time_score: float,
        day_score: float,
        price_score: float,
    ) -> str:
        if court_time_score >= 0.6:
            return f"Bạn thường chơi {slot['courtName']} trong khung giờ {slot['label']}."

        if court_score >= 0.65 and time_score >= 0.45:
            return (
                f"{slot['courtName']} là sân bạn đặt thường xuyên và khung giờ "
                f"{slot['label']} khá sát với thói quen hiện tại."
            )

        if time_score >= 0.65:
            return f"Khung giờ {slot['label']} phù hợp với lịch chơi quen thuộc của bạn."

        if has_history and day_score >= 0.6:
            return f"Bạn thường chơi vào những ngày có lịch tương tự, nên {slot['label']} là lựa chọn hợp lý."

        if price_score >= 0.7:
            return f"Mức giá của khung giờ {slot['label']} gần với những lần đặt gần đây của bạn."

        return f"Khung giờ {slot['label']} đang là lựa chọn cân bằng để đặt nhanh hôm nay."

    def _score_slot(
        self,
        *,
        slot: dict[str, Any],
        available_slots: list[dict[str, Any]],
        history_stats: dict[str, Any],
        target_day: str,
        has_history: bool,
    ) -> dict[str, Any]:
        court_time_score = _normalized_counter_value(
            history_stats["court_time_counts"],
            _build_court_time_key(slot["courtId"], slot["startTime"], slot["endTime"]),
        )
        court_score = _normalized_counter_value(history_stats["court_counts"], slot["courtId"])
        time_score = _normalized_counter_value(
            history_stats["time_counts"],
            _build_time_key(slot["startTime"], slot["endTime"]),
        )
        day_score = _normalized_counter_value(history_stats["day_counts"], target_day)
        price_score = self._score_price_fit(slot["priceVnd"], history_stats["average_price"], available_slots)
        global_score = self._score_global_signal(slot, target_day)

        if has_history:
            total_score = (
                0.45 * court_time_score
                + 0.20 * court_score
                + 0.15 * time_score
                + 0.10 * day_score
                + 0.05 * price_score
                + 0.05 * global_score
            )
        else:
            total_score = 0.50 * global_score + 0.25 * _score_prime_time(slot["startTime"]) + 0.25 * price_score

        return {
            **slot,
            "score": round(float(total_score), 4),
            "reason": self._build_reason(
                slot,
                has_history=has_history,
                court_time_score=court_time_score,
                court_score=court_score,
                time_score=time_score,
                day_score=day_score,
                price_score=price_score,
            ),
        }

File: backend/src/services/recommendation.service.js
const { env } = require('../config/env');
const { assertISODate, formatTimeHHmm, normalizeISODateOnly } = require('../utils/date-time');
const bookingRepository = require('../repositories/booking.repository');
const slotRepository = require('../repositories/slot.repository');

const AI_TIMEOUT_MS = 1200;
const HISTORY_LIMIT = 50;
const MAX_RECOMMENDATIONS = 3;
const WEEKDAY_FORMATTER = new Intl.DateTimeFormat('en-US', {
  weekday: 'long',
  timeZone: 'UTC'
});

function getDayOfWeek(dateString) {
  const date = new Date(`${dateString}T00:00:00.000Z`);
  return Number.isNaN(date.getTime()) ? '' : WEEKDAY_FORMATTER.format(date);
}

function buildTimeLabel(startTime, endTime) {
  return `${formatTimeHHmm(startTime)} - ${formatTimeHHmm(endTime)}`;
}

function buildTimeKey(startTime, endTime) {
  return `${formatTimeHHmm(startTime)}|${formatTimeHHmm(endTime)}`;
}

function buildCourtTimeKey(courtId, startTime, endTime) {
  return `${Number(courtId)}|${buildTimeKey(startTime, endTime)}`;
}

function incrementCounter(map, key) {
  map.set(key, (map.get(key) || 0) + 1);
}

function getNormalizedCount(map, key) {
  if (!map.size) {
    return 0;
  }

  const maxValue = Math.max(...map.values());
  if (maxValue <= 0) {
    return 0;
  }

  return (map.get(key) || 0) / maxValue;
}

function clampScore(value) {
  if (!Number.isFinite(value)) {
    return 0;
  }

  if (value < 0) {
    return 0;
  }

  if (value > 1) {
    return 1;
  }

  return value;
}

function mapAvailableSlot(slot) {
  return {
    courtId: Number(slot.court_id),
    courtName: slot.court_name,
    location: slot.court_location,
    slotId: Number(slot.slot_id),
    label: buildTimeLabel(slot.start_time, slot.end_time),
    startTime: formatTimeHHmm(slot.start_time),
    endTime: formatTimeHHmm(slot.end_time),
    priceVnd: Number(slot.price_vnd)
  };
}

function mapHistoryEntry(entry) {
  return {
    bookingDate: normalizeISODateOnly(entry.booking_date),
    dayOfWeek: getDayOfWeek(normalizeISODateOnly(entry.booking_date)),
    courtId: Number(entry.court_id),
    courtName: entry.court_name,
    startTime: formatTimeHHmm(entry.start_time),
    endTime: formatTimeHHmm(entry.end_time),
    priceVnd: Number(entry.price_vnd ?? entry.amount_vnd ?? 0)
  };
}

function buildHistoryStats(history) {
  const courtTimeCounts = new Map();
  const courtCounts = new Map();
  const timeCounts = new Map();
  const dayCounts = new Map();
  const prices = [];

  for (const item of history) {
    incrementCounter(courtTimeCounts, buildCourtTimeKey(item.courtId, item.startTime, item.endTime));
    incrementCounter(courtCounts, Number(item.courtId));
    incrementCounter(timeCounts, buildTimeKey(item.startTime, item.endTime));
    incrementCounter(dayCounts, item.dayOfWeek);

    if (Number.isFinite(item.priceVnd)) {
      prices.push(Number(item.priceVnd));
    }
  }

  return {
    courtTimeCounts,
    courtCounts,
    timeCounts,
    dayCounts,
    averagePriceVnd: prices.length ? prices.reduce((sum, price) => sum + price, 0) / prices.length : null
  };
}

function scorePriceFit(priceVnd, averagePriceVnd, availableSlots) {
  if (averagePriceVnd && averagePriceVnd > 0) {
    return clampScore(1 - Math.min(Math.abs(priceVnd - averagePriceVnd) / averagePriceVnd, 1));
  }

  const prices = availableSlots.map((slot) => slot.priceVnd);
  const maxPrice = Math.max(...prices);
  const minPrice = Math.min(...prices);
  if (maxPrice === minPrice) {
    return 0.7;
  }

  return clampScore(1 - (priceVnd - minPrice) / (maxPrice - minPrice));
}

function scorePrimeTime(startTime) {
  const hour = Number.parseInt(formatTimeHHmm(startTime).slice(0, 2), 10);
  if (Number.isNaN(hour)) {
    return 0.4;
  }

  return clampScore(1 - Math.min(Math.abs(hour - 19) / 12, 1));
}

function buildReason(slot, { courtTimeScore, courtTimeCount, courtScore, timeScore, dayScore, priceScore, hasHistory }) {
  if (courtTimeScore >= 0.6 && courtTimeCount > 0) {
    return `Bạn thường chơi ${slot.courtName} trong khung giờ ${slot.label} ${courtTimeCount} lần.`;
  }

  if (courtScore >= 0.65 && timeScore >= 0.45) {
    return `${slot.courtName} là sân bạn đặt thường xuyên và khung giờ ${slot.label} khá sát thói quen hiện tại.`;
  }

  if (timeScore >= 0.65) {
    return `Khung giờ ${slot.label} phù hợp với lịch chơi quen thuộc của bạn.`;
  }

  if (hasHistory && dayScore >= 0.6) {
    return `Bạn thường chơi vào ${slot.label} trong ngày có lịch tương tự.`;
  }

  if (priceScore >= 0.7) {
    return `Mức giá của khung giờ ${slot.label} khá gần với những lần đặt gần đây của bạn.`;
  }

  return `Khung giờ ${slot.label} đang là lựa chọn phù hợp để đặt nhanh trong ngày này.`;
}

function scoreFallbackOption(slot, historyStats, targetDay, availableSlots) {
  const courtTimeKey = buildCourtTimeKey(slot.courtId, slot.startTime, slot.endTime);
  const courtTimeScore = getNormalizedCount(
    historyStats.courtTimeCounts,
    courtTimeKey
  );
  const courtTimeCount = historyStats.courtTimeCounts.get(courtTimeKey) || 0;
  const courtScore = getNormalizedCount(historyStats.courtCounts, slot.courtId);
  const timeScore = getNormalizedCount(historyStats.timeCounts, buildTimeKey(slot.startTime, slot.endTime));
  const dayScore = getNormalizedCount(historyStats.dayCounts, targetDay);
  const priceScore = scorePriceFit(slot.priceVnd, historyStats.averagePriceVnd, availableSlots);
  const hasHistory = historyStats.courtCounts.size > 0;

  const totalScore = hasHistory
    ? 0.45 * courtTimeScore + 0.25 * courtScore + 0.15 * timeScore + 0.1 * dayScore + 0.05 * priceScore
    : 0.55 * scorePrimeTime(slot.startTime) + 0.45 * priceScore;

  return {
    ...slot,
    score: Number(totalScore.toFixed(4)),
    reason: hasHistory
      ? buildReason(slot, { courtTimeScore, courtTimeCount, courtScore, timeScore, dayScore, priceScore, hasHistory })
      : `Khung giờ ${slot.label} dễ tiếp cận cho người chơi mới và còn trống để đặt ngay.`
  };
}

function buildFallbackRecommendations(availableSlots, history, targetDate) {
  const targetDay = getDayOfWeek(targetDate);
  const historyStats = buildHistoryStats(history);

  return availableSlots
    .map((slot) => scoreFallbackOption(slot, historyStats, targetDay, availableSlots))
    .sort((left, right) => {
      if (right.score !== left.score) {
        return right.score - left.score;
      }

      if (left.priceVnd !== right.priceVnd) {
        return left.priceVnd - right.priceVnd;
      }

      return left.courtName.localeCompare(right.courtName);
    })
    .slice(0, MAX_RECOMMENDATIONS);
}

function normalizeRecommendedOption(option, availableSlotMap) {
  const slotId = Number(option?.slotId);
  const availableSlot = availableSlotMap.get(slotId);
  if (!availableSlot) {
    return null;
  }

  return {
    ...availableSlot,
    score: Number(Number(option?.score ?? 0).toFixed(4)),
    reason: String(option?.reason || `Khung giờ ${availableSlot.label} phù hợp với lịch chơi của bạn.`).trim()
  };
}

async function fetchRankedRecommendations(payload, availableSlots) {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), AI_TIMEOUT_MS);

  try {
    const response = await fetch(`${env.AI_SERVICE_BASE_URL}/ai/recommendations/score`, {
      method: 'POST',
      signal: controller.signal,
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(payload)
    });

    if (!response.ok) {
      return {
        recommendedOptions: [],
        status: 'unavailable',
        strategy: 'fallback'
      };
    }

    const rawPayload = await response.json();
    const availableSlotMap = new Map(availableSlots.map((slot) => [slot.slotId, slot]));
    const recommendedOptions = Array.isArray(rawPayload?.recommendedOptions)
      ? rawPayload.recommendedOptions
          .map((option) => normalizeRecommendedOption(option, availableSlotMap))
          .filter(Boolean)
          .slice(0, MAX_RECOMMENDATIONS)
      : [];

    return {
      recommendedOptions,
      status: 'ok',
      strategy: String(rawPayload?.strategy || 'personalized')
    };
  } catch (_error) {
    return {
      recommendedOptions: [],
      status: 'unavailable',
      strategy: 'fallback'
    };
  } finally {
    clearTimeout(timeout);
  }
}

function buildResponse(date, recommendedOptions, aiStatus, strategy) {
  const recommendedCourtIds = [...new Set(recommendedOptions.map((option) => option.courtId))];

  return {
    date,
    strategy,
    aiStatus,
    recommendedCourtIds,
    recommendedOptions
  };
}

async function getRecommendedCourts(currentUser, { date }) {
  const bookingDate = String(date || '').trim();
  assertISODate(bookingDate, 'date');

  const rawSlots = await slotRepository.listAvailableSlotsByDate(bookingDate);
  const availableSlots = rawSlots
    .filter((slot) => slot.status === 'AVAILABLE')
    .map(mapAvailableSlot);

  if (availableSlots.length === 0) {
    return buildResponse(bookingDate, [], 'ok', 'empty');
  }

  const history = (await bookingRepository.listRecommendationHistoryByUserId(currentUser.id, { limit: HISTORY_LIMIT })).map(
    mapHistoryEntry
  );

  const aiPayload = {
    user: {
      id: Number(currentUser.id)
    },
    targetDate: bookingDate,
    availableSlots,
    history
  };

  const aiResult = await fetchRankedRecommendations(aiPayload, availableSlots);
  if (aiResult.status === 'ok' && aiResult.recommendedOptions.length > 0) {
    return buildResponse(bookingDate, aiResult.recommendedOptions, aiResult.status, aiResult.strategy);
  }

  const fallbackOptions = buildFallbackRecommendations(availableSlots, history, bookingDate);
  return buildResponse(bookingDate, fallbackOptions, aiResult.status, aiResult.strategy);
}

module.exports = {
  getRecommendedCourts
};
